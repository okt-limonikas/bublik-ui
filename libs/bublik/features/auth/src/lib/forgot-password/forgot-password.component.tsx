/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { forwardRef, useImperativeHandle } from 'react';
import { useForm, UseFormReturn } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';

import {
	ButtonTw,
	FormAlertError,
	TextField
} from '@/shared/tailwind-ui';
import {
	ForgotPasswordInputs,
	ForgotPasswordSchema
} from '@/shared/types';

import { AuthFormLayout } from '../auth-form-layout';

export type ForgotPasswordFormHandle = UseFormReturn<ForgotPasswordInputs>;

export type ForgotPasswordFormProps = {
	onSubmit: (form: ForgotPasswordInputs) => void;
	defaultValues?: ForgotPasswordInputs;
};

export const ForgotPasswordForm = forwardRef<
	ForgotPasswordFormHandle,
	ForgotPasswordFormProps
>((props, ref) => {
	const { onSubmit, defaultValues = { email: '' } } = props;

	const form = useForm<ForgotPasswordInputs>({
		defaultValues,
		resolver: zodResolver(ForgotPasswordSchema)
	});

	useImperativeHandle(ref, () => form, [form]);

	const rootError = form.formState.errors.root;

	return (
		<AuthFormLayout
			label="Forgot Password"
			description="If you forgot your password we will send you an email with link to reset it"
		>
			{rootError ? (
				<div className="mb-6">
					<FormAlertError title={'Error'} description={rootError.message} />
				</div>
			) : null}
			<form
				onSubmit={form.handleSubmit((form) => onSubmit(form))}
				className="flex flex-col gap-6"
			>
				<TextField
					name="email"
					label="Email"
					placeholder="Email"
					control={form.control}
				/>
				<ButtonTw type="submit" disabled={form.formState.isSubmitting}>
					Send Link
				</ButtonTw>
			</form>
		</AuthFormLayout>
	);
});
