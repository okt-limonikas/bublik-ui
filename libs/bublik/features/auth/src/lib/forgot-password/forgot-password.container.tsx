/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { useRef } from 'react';
import { useNavigate } from 'react-router-dom';

import { ForgotPasswordInputs } from '@/shared/types';
import { useRequestResetPasswordMutation } from '@/services/bublik-api';
import { toast } from '@/shared/tailwind-ui';
import { setErrorsOnForm } from '@/shared/utils';

import {
	ForgotPasswordForm,
	ForgotPasswordFormHandle
} from './forgot-password.component';

export const ForgotPasswordFormContainer = () => {
	const navigate = useNavigate();
	const [requestPasswordReset] = useRequestResetPasswordMutation();
	const formRef = useRef<ForgotPasswordFormHandle>(null);

	const handleSubmit = async (form: ForgotPasswordInputs) => {
		const formHandle = formRef.current;

		if (!formHandle) return;

		try {
			await requestPasswordReset(form).unwrap();

			toast.success('Email with link to reset password has been sent!');
			navigate('/auth/login');
		} catch (e: unknown) {
			setErrorsOnForm(e, { handle: formHandle });
			toast.error('Failed to send reset password email!');
		}
	};

	return <ForgotPasswordForm onSubmit={handleSubmit} ref={formRef} />;
};
