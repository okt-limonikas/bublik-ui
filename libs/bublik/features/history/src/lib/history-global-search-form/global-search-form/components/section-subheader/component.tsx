/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { FC } from 'react';

export interface FormSectionSubheaderProps {
	name: string;
}

export const FormSectionSubheader: FC<FormSectionSubheaderProps> = ({
	name
}) => {
	return (
		<div className="mb-2">
			<span className="text-[0.6875rem] leading-[0.875rem]">{name}</span>
		</div>
	);
};
