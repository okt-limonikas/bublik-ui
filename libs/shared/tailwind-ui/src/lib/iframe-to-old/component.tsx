/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { ComponentPropsWithRef, FC } from 'react';

import { cn } from '../utils';

export type IframeToOldBublikProps = ComponentPropsWithRef<'iframe'>;

export const IframeToOldBublik: FC<IframeToOldBublikProps> = (props) => {
	return (
		<iframe
			className={cn('w-full h-full border-none', props.className)}
			title="Old Bublik"
			{...props}
		/>
	);
};
