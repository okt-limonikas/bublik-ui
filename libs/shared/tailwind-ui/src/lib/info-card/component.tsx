/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { ComponentPropsWithRef, forwardRef } from 'react';

export interface InfoCardProps extends ComponentPropsWithRef<'div'> {
	header: string;
	message: string;
}

export const InfoCard = forwardRef<HTMLDivElement, InfoCardProps>(
	(props, ref) => {
		const { header, message } = props;

		return (
			<div
				className="flex items-center justify-center w-full h-screen"
				data-testid="info-card"
			>
				<div
					className="bg-white min-w-[600px] max-w-[768px] flex flex-col rounded-2xl shadow-2xl"
					ref={ref}
				>
					<div className="p-4 border-b border-border-primary">
						<h2>{header}</h2>
					</div>
					<div className="flex-1 p-4">
						<p>{message}</p>
					</div>
				</div>
			</div>
		);
	}
);
