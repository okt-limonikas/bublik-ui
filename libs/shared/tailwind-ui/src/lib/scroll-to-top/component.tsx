/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { ComponentPropsWithRef, FC, forwardRef } from 'react';
import { AnimatePresence, motion, Variants } from 'framer-motion';

import { useScrollToTopPage } from '@/shared/hooks';

import { Icon } from '../icon';

const variants: Variants = {
	visible: { y: 0, opacity: 1 },
	hidden: { y: 72, opacity: 0 }
};

export interface ScrollToTopProps extends ComponentPropsWithRef<'button'> {
	offset?: number;
}

export const ScrollToTopPage: FC<ScrollToTopProps> = forwardRef(
	({ offset = 0 }, ref) => {
		const { isVisible, scrollToTop } = useScrollToTopPage(offset);

		return (
			<AnimatePresence>
				{isVisible && (
					<motion.div
						variants={variants}
						initial="hidden"
						animate="visible"
						exit="hidden"
						transition={{ type: 'just' }}
						className="fixed z-40 bottom-4 right-4"
					>
						<button
							className="flex items-center gap-1 p-3 text-white rounded-full bg-primary transition-all text-[1rem] font-medium leading-[0.875rem] border-2 border-transparent hover:border-[#94b0ff]"
							onClick={scrollToTop}
							data-testid="scroll-to-top"
							aria-label="Scroll to top"
							ref={ref}
						>
							<Icon name="ArrowShortTop" className="grid place-items-center" />
						</button>
					</motion.div>
				)}
			</AnimatePresence>
		);
	}
);
