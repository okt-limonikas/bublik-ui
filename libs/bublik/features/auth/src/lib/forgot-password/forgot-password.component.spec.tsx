/* SPDX-License-Identifier: Apache-2.0 */
/* SPDX-FileCopyrightText: 2021-2023 OKTET Labs Ltd. */
import { describe, it, expect, vi } from 'vitest';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ForgotPasswordForm } from './forgot-password.component';

describe('ForgotPasswordForm', () => {
	it('should match snapshot', async () => {
		const onSubmit = vi.fn();

		const { asFragment } = render(<ForgotPasswordForm onSubmit={onSubmit} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should call onSubmit with the email when the form is submitted', async () => {
		const onSubmit = vi.fn();

		const { getByLabelText, getByRole } = render(
			<ForgotPasswordForm onSubmit={onSubmit} />
		);

		await userEvent.type(getByLabelText(/email/i), 'test@example.com');
		await userEvent.click(getByRole('button'));

		expect(onSubmit).toHaveBeenCalledWith({
			email: 'test@example.com'
		});
	});

	it('should dispaly error when email is not correct', async () => {
		const onSubmit = vi.fn();

		const { getByLabelText, asFragment, getByRole } = render(
			<ForgotPasswordForm onSubmit={onSubmit} />
		);

		await userEvent.type(getByLabelText(/email/i), 'testexample.com');
		await userEvent.click(getByRole('button'));

		expect(asFragment()).toMatchSnapshot();

		expect(onSubmit).not.toHaveBeenCalled();
	});
});
