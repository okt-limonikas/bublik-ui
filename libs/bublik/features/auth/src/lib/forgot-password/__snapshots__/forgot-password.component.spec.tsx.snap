// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ForgotPasswordForm > should dispaly error when email is not correct 1`] = `
<DocumentFragment>
  <div
    class="absolute z-0 rounded-full w-96 h-96 blur-3xl mix-blend-multiply opacity-50 animate-auth-blob bg-primary left-[10%] bottom-[15%]"
  />
  <div
    class="absolute z-0 rounded-full w-96 h-96 blur-3xl mix-blend-multiply opacity-50 animate-auth-blob bg-bg-warning right-[25%] top-[10%]"
    style="animation-delay: 2s;"
  />
  <div
    class="absolute z-0 rounded-full w-96 h-96 blur-3xl mix-blend-multiply opacity-50 animate-auth-blob bg-bg-error bottom-[10%] right-[17%]"
    style="animation-delay: 4s;"
  />
  <div
    class="w-full sm:max-w-md p-6 bg-white sm:rounded-lg md:shadow min-w-[420px] z-10 relative"
  >
    <div
      class="flex flex-col gap-2 mb-6"
    >
      <h1
        class="text-2xl font-bold leading-tight tracking-tight text-text-primary"
      >
        Forgot Password
      </h1>
      <p
        class="text-sm font-light"
      >
        If you forgot your password we will send you an email with link to reset it
      </p>
    </div>
    <form
      class="flex flex-col gap-6"
    >
      <div
        class="relative"
      >
        <div
          class="flex items-center gap-3"
        >
          <label
            class="font-normal text-text-secondary text-[0.875rem] absolute top-[-11px] left-2 bg-white"
            data-testid="input-label"
            for="email"
          >
            Email
          </label>
          <input
            autocomplete="off"
            class="w-full px-3.5 py-[7px] outline-none border rounded text-text-secondary transition-all disabled:text-text-menu disabled:cursor-not-allowed focus:ring-transparent border-bg-error hover:border-bg-error caret-bg-error shadow-text-field-error focus:border-bg-error active:shadow-none focus:shadow-text-field-error"
            data-testid="input"
            id="email"
            name="email"
            placeholder="Email"
            spellcheck="false"
            type="text"
            value="testexample.com"
          />
          <svg
            class="grid place-items-center"
            fill="none"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M16 8C16 12.4183 12.4183 16 8 16C3.58172 16 0 12.4183 0 8C0 3.58172 3.58172 0 8 0C12.4183 0 16 3.58172 16 8ZM7.99463 12.546C7.58063 12.546 7.24463 12.21 7.24463 11.796V7.37598C7.24463 6.96198 7.58063 6.62598 7.99463 6.62598C8.40863 6.62598 8.74463 6.96198 8.74463 7.37598V11.796C8.74463 12.21 8.40863 12.546 7.99463 12.546ZM7.25 4.2041C7.25 4.6181 7.581 4.9541 7.995 4.9541H8.005C8.419 4.9541 8.755 4.6181 8.755 4.2041C8.755 3.7901 8.419 3.4541 8.005 3.4541C7.591 3.4541 7.25 3.7901 7.25 4.2041Z"
              fill="#F95C78"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <span
          class="text-[0.75rem] font-normal text-bg-error"
          data-testid="input-error-message"
        >
          Must be a valid email
        </span>
      </div>
      <button
        class="inline-flex items-center justify-center focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary transition-all appearance-none select-none bg-primary text-white disabled:shadow-[inset_0_0_0_1px_hsl(var(--colors-border-primary))] disabled:bg-white disabled:hover:bg-white disabled:text-text-menu px-3 text-[0.875rem] leading-[1.5rem] rounded-md hover:shadow-[inset_0_0_0_3px_#94b0ff] py-2"
        type="submit"
      >
        Send Link
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`ForgotPasswordForm > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="absolute z-0 rounded-full w-96 h-96 blur-3xl mix-blend-multiply opacity-50 animate-auth-blob bg-primary left-[10%] bottom-[15%]"
  />
  <div
    class="absolute z-0 rounded-full w-96 h-96 blur-3xl mix-blend-multiply opacity-50 animate-auth-blob bg-bg-warning right-[25%] top-[10%]"
    style="animation-delay: 2s;"
  />
  <div
    class="absolute z-0 rounded-full w-96 h-96 blur-3xl mix-blend-multiply opacity-50 animate-auth-blob bg-bg-error bottom-[10%] right-[17%]"
    style="animation-delay: 4s;"
  />
  <div
    class="w-full sm:max-w-md p-6 bg-white sm:rounded-lg md:shadow min-w-[420px] z-10 relative"
  >
    <div
      class="flex flex-col gap-2 mb-6"
    >
      <h1
        class="text-2xl font-bold leading-tight tracking-tight text-text-primary"
      >
        Forgot Password
      </h1>
      <p
        class="text-sm font-light"
      >
        If you forgot your password we will send you an email with link to reset it
      </p>
    </div>
    <form
      class="flex flex-col gap-6"
    >
      <div
        class="relative"
      >
        <div
          class="flex items-center gap-3"
        >
          <label
            class="font-normal text-text-secondary text-[0.875rem] absolute top-[-11px] left-2 bg-white"
            data-testid="input-label"
            for="email"
          >
            Email
          </label>
          <input
            autocomplete="off"
            class="w-full px-3.5 py-[7px] outline-none border border-border-primary rounded text-text-secondary transition-all hover:border-primary disabled:text-text-menu disabled:cursor-not-allowed focus:border-primary focus:shadow-text-field active:shadow-none focus:ring-transparent"
            data-testid="input"
            id="email"
            name="email"
            placeholder="Email"
            spellcheck="false"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="inline-flex items-center justify-center focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary transition-all appearance-none select-none bg-primary text-white disabled:shadow-[inset_0_0_0_1px_hsl(var(--colors-border-primary))] disabled:bg-white disabled:hover:bg-white disabled:text-text-menu px-3 text-[0.875rem] leading-[1.5rem] rounded-md hover:shadow-[inset_0_0_0_3px_#94b0ff] py-2"
        type="submit"
      >
        Send Link
      </button>
    </form>
  </div>
</DocumentFragment>
`;
